File: Go/webApi/utils/callbackcrypto.go
package utils

import (
	"bytes"
	"crypto/aes"
	"crypto/cipher"
	"crypto/sha1"
	"encoding/base64"
	"errors"
	"fmt"
	"math/rand"
	"sort"
	"strconv"
	"strings"
	"time"
)

// CallbackCrypto handles encryption and decryption for callback messages
type CallbackCrypto struct {
	aesKey   []byte
	clientID string
}

const randomLength = 16

// NewCallbackCrypto initializes a new CallbackCrypto instance
func NewCallbackCrypto(encodingAesKey, clientID string) (*CallbackCrypto, error) {
	if encodingAesKey == "" {
		return nil, errors.New("encryption key cannot be empty")
	}
	aesKey, err := base64.StdEncoding.DecodeString(encodingAesKey)
	if err != nil {
		return nil, fmt.Errorf("invalid base64 encodingAesKey: %v", err)
	}
	return &CallbackCrypto{
		aesKey:   aesKey,
		clientID: clientID,
	}, nil
}

// GetEncryptedMap encrypts a plaintext message and returns a map with encrypted data
func (c *CallbackCrypto) GetEncryptedMap(plaintext string, timestamp int64) (map[string]string, error) {
	if plaintext == "" {
		return nil, errors.New("plaintext cannot be empty")
	}

	nonce := getRandomStr(randomLength)
	encrypt, err := c.encrypt(nonce, plaintext)
	if err != nil {
		return nil, err
	}
	signature := c.getSignature(strconv.FormatInt(timestamp, 10), nonce, encrypt)

	return map[string]string{
		"signature": signature,
		"encrypt":   encrypt,
		"timestamp": strconv.FormatInt(timestamp, 10),
		"nonce":     nonce,
	}, nil
}

// GetDecryptMsg decrypts an encrypted message after verifying the signature
func (c *CallbackCrypto) GetDecryptMsg(msgSignature, timestamp, nonce, encryptMsg string) (string, error) {
	signature := c.getSignature(timestamp, nonce, encryptMsg)
	if signature != msgSignature {
		return "", errors.New("signature mismatch")
	}
	return c.decrypt(encryptMsg)
}

// encrypt encrypts a plaintext message
func (c *CallbackCrypto) encrypt(nonce, plaintext string) (string, error) {
	randomBytes := []byte(nonce)
	plainTextBytes := []byte(plaintext)
	lengthBytes := int2Bytes(len(plainTextBytes))
	clientIDBytes := []byte(c.clientID)

	data := bytes.Join([][]byte{randomBytes, lengthBytes, plainTextBytes, clientIDBytes}, nil)
	paddedData, err := pkcs7Pad(data, aes.BlockSize)
	if err != nil {
		return "", err
	}

	block, err := aes.NewCipher(c.aesKey)
	if err != nil {
		return "", err
	}
	iv := c.aesKey[:16]
	ciphertext := make([]byte, len(paddedData))
	mode := cipher.NewCBCEncrypter(block, iv)
	mode.CryptBlocks(ciphertext, paddedData)

	return base64.StdEncoding.EncodeToString(ciphertext), nil
}

// decrypt decrypts an encrypted message
func (c *CallbackCrypto) decrypt(text string) (string, error) {
	ciphertext, err := base64.StdEncoding.DecodeString(text)
	if err != nil {
		return "", fmt.Errorf("invalid base64 ciphertext: %v", err)
	}

	block, err := aes.NewCipher(c.aesKey)
	if err != nil {
		return "", err
	}
	iv := c.aesKey[:16]
	if len(ciphertext)%aes.BlockSize != 0 {
		return "", errors.New("ciphertext is not a multiple of the block size")
	}

	plaintext := make([]byte, len(ciphertext))
	mode := cipher.NewCBCDecrypter(block, iv)
	mode.CryptBlocks(plaintext, ciphertext)

	unpadded, err := pkcs7Unpad(plaintext)
	if err != nil {
		return "", err
	}

	if len(unpadded) < randomLength+4 {
		return "", errors.New("decrypted data too short")
	}

	plainTextLength := bytes2Int(unpadded[randomLength : randomLength+4])
	if len(unpadded) < randomLength+4+plainTextLength {
		return "", errors.New("invalid plaintext length")
	}

	plaintextData := unpadded[randomLength+4 : randomLength+4+plainTextLength]
	msgClientID := string(unpadded[randomLength+4+plainTextLength:])
	if msgClientID != c.clientID {
		return "", errors.New("clientID mismatch in decrypted message")
	}

	return string(plaintextData), nil
}

// getSignature generates a SHA1 signature
func (c *CallbackCrypto) getSignature(timestamp, nonce, encrypt string) string {
	array := []string{c.clientID, timestamp, nonce, encrypt}
	sort.Strings(array)
	data := strings.Join(array, "")
	hash := sha1.Sum([]byte(data))
	return fmt.Sprintf("%x", hash)
}

// Utility functions

func getRandomStr(count int) string {
	chars := "ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789"
	sb := strings.Builder{}
	rand.Seed(time.Now().UnixNano())
	for i := 0; i < count; i++ {
		sb.WriteByte(chars[rand.Intn(len(chars))])
	}
	return sb.String()
}

func int2Bytes(n int) []byte {
	return []byte{
		byte(n >> 24 & 0xFF),
		byte(n >> 16 & 0xFF),
		byte(n >> 8 & 0xFF),
		byte(n & 0xFF),
	}
}

func bytes2Int(b []byte) int {
	return int(b[0])<<24 | int(b[1])<<16 | int(b[2])<<8 | int(b[3])
}

func pkcs7Pad(data []byte, blockSize int) ([]byte, error) {
	if blockSize <= 0 {
		return nil, errors.New("invalid block size")
	}
	padding := blockSize - len(data)%blockSize
	padText := bytes.Repeat([]byte{byte(padding)}, padding)
	return append(data, padText...), nil
}

func pkcs7Unpad(data []byte) ([]byte, error) {
	if len(data) == 0 {
		return nil, errors.New("empty data")
	}
	padding := int(data[len(data)-1])
	if padding < 1 || padding > len(data) {
		return nil, errors.New("invalid padding")
	}
	return data[:len(data)-padding], nil
}

File: Nodejs/console/callbackCrypto.js
const crypto = require('crypto');

class CallbackCrypto {
    static RANDOM_LENGTH = 16;

    constructor(encodingAesKey, clientId) {
        if (!encodingAesKey) {
            throw new Error('Encryption key cannot be empty');
        }
        this._aesKey = Buffer.from(encodingAesKey, 'base64');
        if (this._aesKey.length !== 16) {
            throw new Error(`Invalid AES key length: expected 16 bytes, got ${this._aesKey.length} bytes`);
        }
        this._clientId = clientId;
    }

    getEncryptedMap(plaintext, timestamp = Date.now()) {
        if (!plaintext) {
            throw new Error('Plaintext cannot be empty');
        }
        const nonce = Utils.getRandomStr(CallbackCrypto.RANDOM_LENGTH);
        const encrypt = this.encrypt(nonce, plaintext);
        const signature = this.getSignature(timestamp.toString(), nonce, encrypt);

        return {
            signature,
            encrypt,
            timestamp: timestamp.toString(),
            nonce
        };
    }

    getDecryptMsg(msgSignature, timestamp, nonce, encryptMsg) {
        const signature = this.getSignature(timestamp, nonce, encryptMsg);
        if (signature !== msgSignature) {
            throw new Error('Signature mismatch');
        }
        return this.decrypt(encryptMsg);
    }

    encrypt(nonce, plaintext) {
        const randomBytes = Buffer.from(nonce, 'utf8');
        const plainTextBytes = Buffer.from(plaintext, 'utf8');
        const lengthBytes = Utils.int2Bytes(plainTextBytes.length);
        const clientIdBytes = Buffer.from(this._clientId, 'utf8');

        let data = Buffer.concat([randomBytes, lengthBytes, plainTextBytes, clientIdBytes]);
        const paddedData = PKCS7Padding.getPaddingBytes(data);

        const cipher = crypto.createCipheriv('aes-128-cbc', this._aesKey, this._aesKey.slice(0, 16));
        cipher.setAutoPadding(false);
        let encrypted = cipher.update(paddedData);
        encrypted = Buffer.concat([encrypted, cipher.final()]);

        return encrypted.toString('base64');
    }

    decrypt(text) {
        const decipher = crypto.createDecipheriv('aes-128-cbc', this._aesKey, this._aesKey.slice(0, 16));
        decipher.setAutoPadding(false);
        let decrypted = decipher.update(Buffer.from(text, 'base64'));
        decrypted = Buffer.concat([decrypted, decipher.final()]);

        const unpadded = PKCS7Padding.removePaddingBytes(decrypted);
        const plainTextLength = Utils.bytes2Int(unpadded.slice(CallbackCrypto.RANDOM_LENGTH, CallbackCrypto.RANDOM_LENGTH + 4));
        const plaintext = unpadded.slice(CallbackCrypto.RANDOM_LENGTH + 4, CallbackCrypto.RANDOM_LENGTH + 4 + plainTextLength).toString('utf8');
        const msgClientId = unpadded.slice(CallbackCrypto.RANDOM_LENGTH + 4 + plainTextLength).toString('utf8');

        if (msgClientId !== this._clientId) {
            throw new Error('ClientID mismatch in decrypted message');
        }
        return plaintext;
    }

    getSignature(timestamp, nonce, encrypt) {
        const array = [this._clientId, timestamp, nonce, encrypt];
        array.sort();
        const data = array.join('');
        return crypto.createHash('sha1').update(data, 'ascii').digest('hex');
    }
}

class Utils {
    static getRandomStr(count) {
        const chars = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789';
        let result = '';
        for (let i = 0; i < count; i++) {
            result += chars.charAt(Math.floor(Math.random() * chars.length));
        }
        return result;
    }

    static int2Bytes(number) {
        const buffer = Buffer.alloc(4);
        buffer.writeInt32BE(number, 0);
        return buffer;
    }

    static bytes2Int(byteArr) {
        return byteArr.readInt32BE(0);
    }
}

class PKCS7Padding {
    static BLOCK_SIZE = 32;

    static getPaddingBytes(data) {
        const amountToPad = this.BLOCK_SIZE - (data.length % this.BLOCK_SIZE) || this.BLOCK_SIZE;
        const padChr = String.fromCharCode(amountToPad);
        return Buffer.concat([data, Buffer.alloc(amountToPad, padChr)]);
    }

    static removePaddingBytes(decrypted) {
        const pad = decrypted[decrypted.length - 1];
        if (pad < 1 || pad > this.BLOCK_SIZE) {
            return decrypted;
        }
        return decrypted.slice(0, decrypted.length - pad);
    }
}

module.exports = { CallbackCrypto, Utils, PKCS7Padding };